File: apps/aarf-pocket-android/app/src/main/java/dev/deepiri/aarflingo/data/CollarCbor.kt
package dev.deepiri.aarflingo.data

object CollarGatt {
    const val ADV_NAME = "aarf-collar"
    const val SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
    const val NOTIFY_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"
}

data class CollarVitals(
    val v: Int,
    val tsMs: Long,
    val hrBpm: Int,
    val rmssdMs: Int,
    val vbatV: Float,
    val imuRms: Float,
    val bark: Boolean,
    val ppgOk: Boolean,
    val still: Boolean,
    val pant: Boolean,
    val rrBpm: Int,
    val arousal: Float,
    val fault: String?,
    val source: String,
)

object CollarCbor {
    fun decode(buf: ByteArray): CollarVitals? {
        if (buf.isEmpty() || (buf[0].toInt() and 0xE0) != 0xA0) return null
        val pairs = buf[0].toInt() and 0x1F
        var i = 1
        val map = HashMap<String, Any?>()
        repeat(pairs) {
            val key = text(buf, i) ?: return null
            i = key.second
            val value = value(buf, i) ?: return null
            i = value.second
            map[key.first] = value.first
        }
        return CollarVitals(
            v = (map["v"] as? Int) ?: 0,
            tsMs = ((map["ts_ms"] as? Int) ?: 0).toLong(),
            hrBpm = (map["hr_bpm"] as? Int) ?: 0,
            rmssdMs = (map["rmssd_ms"] as? Int) ?: 0,
            vbatV = (map["vbat_v"] as? Float) ?: 0f,
            imuRms = (map["imu_rms"] as? Float) ?: 0f,
            bark = (map["bark"] as? Boolean) ?: false,
            ppgOk = (map["ppg_ok"] as? Boolean) ?: false,
            still = (map["still"] as? Boolean) ?: false,
            pant = (map["pant"] as? Boolean) ?: false,
            rrBpm = (map["rr_bpm"] as? Int) ?: 0,
            arousal = (map["arousal"] as? Float) ?: 0f,
            fault = map["fault"] as? String,
            source = (map["source"] as? String) ?: "sensors",
        )
    }

    private fun u8(buf: ByteArray, i: Int): Pair<Int, Int>? {
        if (i >= buf.size) return null
        return (buf[i].toInt() and 0xFF) to (i + 1)
    }

    private fun uint(buf: ByteArray, i0: Int): Pair<Int, Int>? {
        val t = u8(buf, i0) ?: return null
        if (t.first shr 5 != 0) return null
        val ai = t.first and 0x1F
        var i = t.second
        if (ai < 24) return ai to i
        if (ai == 24) {
            val v = u8(buf, i) ?: return null
            return v.first to v.second
        }
        if (ai == 25) {
            val hi = u8(buf, i) ?: return null
            val lo = u8(buf, hi.second) ?: return null
            return ((hi.first shl 8) or lo.first) to lo.second
        }
        return null
    }

    private fun text(buf: ByteArray, i0: Int): Pair<String, Int>? {
        val t = u8(buf, i0) ?: return null
        if (t.first shr 5 != 3) return null
        val n = t.first and 0x1F
        val i = t.second
        if (n >= 24 || i + n > buf.size) return null
        return String(buf, i, n, Charsets.UTF_8) to (i + n)
    }

    private fun f32(buf: ByteArray, i0: Int): Pair<Float, Int>? {
        val t = u8(buf, i0) ?: return null
        if (t.first != 0xFA || t.second + 4 > buf.size) return null
        val i = t.second
        val bits = (buf[i].toInt() and 0xFF shl 24) or
            (buf[i + 1].toInt() and 0xFF shl 16) or
            (buf[i + 2].toInt() and 0xFF shl 8) or
            (buf[i + 3].toInt() and 0xFF)
        return Float.fromBits(bits) to (i + 4)
    }

    private fun value(buf: ByteArray, i0: Int): Pair<Any?, Int>? {
        if (i0 >= buf.size) return null
        val t = buf[i0].toInt() and 0xFF
        if (t == 0xF4) return false to (i0 + 1)
        if (t == 0xF5) return true to (i0 + 1)
        if (t == 0xF6) return null to (i0 + 1)
        if (t == 0xFA) return f32(buf, i0)
        if (t shr 5 == 0) return uint(buf, i0)
        if (t shr 5 == 3) return text(buf, i0)
        return null
    }
}

File: apps/aarf-pocket-ios/AarflingoPocket/Models/CollarCbor.swift
import Foundation

enum CollarGatt {
    static let advName = "aarf-collar"
    static let serviceUUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"
    static let notifyUUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"
}

struct CollarVitals: Equatable {
    var v: Int
    var tsMs: UInt32
    var hrBpm: Int
    var rmssdMs: Int
    var vbatV: Double
    var imuRms: Double
    var bark: Bool
    var ppgOk: Bool
    var still: Bool
    var pant: Bool
    var rrBpm: Int
    var arousal: Double
    var gyroRms: Double
    var puckC: Double
    var skinC: Double
    var fault: String?
    var source: String
}

enum CollarCbor {
    static func decode(_ data: Data) -> CollarVitals? {
        let buf = [UInt8](data)
        guard !buf.isEmpty, buf[0] & 0xE0 == 0xA0 else { return nil }
        let pairs = Int(buf[0] & 0x1F)
        var i = 1
        var map: [String: Any] = [:]
        for _ in 0..<pairs {
            guard let key = text(buf, &i), let val = value(buf, &i) else { return nil }
            map[key] = val
        }
        return CollarVitals(
            v: map["v"] as? Int ?? 0,
            tsMs: UInt32(truncatingIfNeeded: (map["ts_ms"] as? Int) ?? 0),
            hrBpm: map["hr_bpm"] as? Int ?? 0,
            rmssdMs: map["rmssd_ms"] as? Int ?? 0,
            vbatV: map["vbat_v"] as? Double ?? 0,
            imuRms: map["imu_rms"] as? Double ?? 0,
            bark: map["bark"] as? Bool ?? false,
            ppgOk: map["ppg_ok"] as? Bool ?? false,
            still: map["still"] as? Bool ?? false,
            pant: map["pant"] as? Bool ?? false,
            rrBpm: map["rr_bpm"] as? Int ?? 0,
            arousal: map["arousal"] as? Double ?? 0,
            gyroRms: map["gyro"] as? Double ?? 0,
            puckC: map["puck_c"] as? Double ?? 0,
            skinC: map["skin_c"] as? Double ?? 0,
            fault: map["fault"] as? String,
            source: map["source"] as? String ?? "sensors"
        )
    }

    private static func u8(_ buf: [UInt8], _ i: inout Int) -> UInt8? {
        guard i < buf.count else { return nil }
        let v = buf[i]
        i += 1
        return v
    }

    private static func uint(_ buf: [UInt8], _ i: inout Int) -> Int? {
        guard let t = u8(buf, &i), t >> 5 == 0 else { return nil }
        let ai = Int(t & 0x1F)
        if ai < 24 { return ai }
        if ai == 24 { return u8(buf, &i).map(Int.init) }
        if ai == 25, let hi = u8(buf, &i), let lo = u8(buf, &i) {
            return (Int(hi) << 8) | Int(lo)
        }
        return nil
    }

    private static func text(_ buf: [UInt8], _ i: inout Int) -> String? {
        guard let t = u8(buf, &i), t >> 5 == 3 else { return nil }
        let n = Int(t & 0x1F)
        guard n < 24, i + n <= buf.count else { return nil }
        let s = String(bytes: buf[i..<(i + n)], encoding: .utf8)
        i += n
        return s
    }

    private static func f32(_ buf: [UInt8], _ i: inout Int) -> Double? {
        guard let t = u8(buf, &i), t == 0xFA, i + 4 <= buf.count else { return nil }
        var u: UInt32 = 0
        u |= UInt32(buf[i]) << 24
        u |= UInt32(buf[i + 1]) << 16
        u |= UInt32(buf[i + 2]) << 8
        u |= UInt32(buf[i + 3])
        i += 4
        return Double(Float(bitPattern: u))
    }

    private static func value(_ buf: [UInt8], _ i: inout Int) -> Any? {
        guard i < buf.count else { return nil }
        let t = buf[i]
        if t == 0xF4 { i += 1; return false }
        if t == 0xF5 { i += 1; return true }
        if t == 0xF6 { i += 1; return nil }
        if t == 0xFA { return f32(buf, &i) }
        if t >> 5 == 0 { return uint(buf, &i) }
        if t >> 5 == 3 { return text(buf, &i) }
        return nil
    }
}
